File: packages/core/validation.ts
import { BOARD_SIZE, MAX_CANDIDATES } from "@pkg/shared/constants";
import type { BoardState, Coordinate } from "@pkg/shared/schemas";

/**
 * Checks if a coordinate is within the board boundaries.
 *
 * @param coord - The coordinate to check
 * @returns true if the coordinate is within [0, BOARD_SIZE) for both x and y
 */
export function isInBounds(coord: Coordinate): boolean {
  return (
    coord.x >= 0 && coord.x < BOARD_SIZE && coord.y >= 0 && coord.y < BOARD_SIZE
  );
}

/**
 * Checks if a cell at the given coordinate is empty.
 *
 * @param board - The current board state
 * @param coord - The coordinate to check
 * @returns true if the cell is empty (null), false if occupied or invalid
 */
export function isEmpty(board: BoardState, coord: Coordinate): boolean {
  const row = board[coord.y];
  if (!row) return false;
  return row[coord.x] === null;
}

/**
 * Checks if a coordinate is a valid candidate for stone placement.
 * A valid candidate must be within bounds and the cell must be empty.
 *
 * @param board - The current board state
 * @param coord - The coordinate to validate
 * @returns true if the coordinate is valid for placing a stone
 */
export function isValidCandidate(
  board: BoardState,
  coord: Coordinate,
): boolean {
  return isInBounds(coord) && isEmpty(board, coord);
}

export type CandidateValidationError =
  | "invalid_candidate_count"
  | "invalid_candidate_position";

export type CandidateValidationResult =
  | { ok: true }
  | { ok: false; error: CandidateValidationError };

/**
 * Validates the candidate list for a turn against count and board positions.
 *
 * @param board - The current board state
 * @param candidates - Candidate coordinates submitted by the player
 * @returns Validation result with error kind when invalid
 */
export function validateCandidates(
  board: BoardState,
  candidates: Coordinate[],
): CandidateValidationResult {
  if (candidates.length < 1 || candidates.length > MAX_CANDIDATES) {
    return { ok: false, error: "invalid_candidate_count" };
  }
  for (const coord of candidates) {
    if (!isValidCandidate(board, coord)) {
      return { ok: false, error: "invalid_candidate_position" };
    }
  }
  return { ok: true };
}

File: ml/src/ungomoku_ml/rules/validation.py
"""Candidate validation. Port of packages/core/validation.ts."""

import numpy as np

from ungomoku_ml.rules.board import EMPTY
from ungomoku_ml.rules.constants import BOARD_SIZE, MAX_CANDIDATES

INVALID_COUNT = "invalid_candidate_count"
INVALID_POSITION = "invalid_candidate_position"


def is_valid_candidate(board: np.ndarray, x: int, y: int) -> bool:
    return 0 <= x < BOARD_SIZE and 0 <= y < BOARD_SIZE and board[y, x] == EMPTY


def validate_candidates(board: np.ndarray, candidates: list[tuple[int, int]]) -> str | None:
    """Returns None when valid, otherwise the error kind (matching the TS union)."""
    if len(candidates) < 1 or len(candidates) > MAX_CANDIDATES:
        return INVALID_COUNT
    for x, y in candidates:
        if not is_valid_candidate(board, x, y):
            return INVALID_POSITION
    return None
